src/qibo/gates/abstract.py

```diff
@@ -359,18 +360,103 @@ def controlled_by(self, *qubits: int) -> "Gate":
             self.control_qubits = qubits
         return self
 
-    def decompose(self, *free) -> List["Gate"]:
+    def _base_decompose(self, *free, use_toffolis=True) -> List["Gate"]:
+        """Base decomposition for gates.
+
+        Returns a list containing the gate itself. Should be overridden by
+        subclasses that support decomposition to simpler gates.
+
+        Args:
+            free: Ids of free qubits to use for the gate decomposition.
+            use_toffolis: If ``True`` the decomposition contains only ``TOFFOLI`` gates.
+                If ``False`` a congruent representation is used for ``TOFFOLI`` gates.
+                See :class:`qibo.gates.TOFFOLI` for more details on this representation.
+
+        Returns:
+            list: Synthesis of the original gate in another gate set.
+        """
+        return [self.__class__(*self.init_args, **self.init_kwargs)]
+
+    @staticmethod
+    def _gates_cancel(g1, g2):
+        """Check if same gate type, same targets, same controls"""
+        if g1.__class__ != g2.__class__:
+            return False
+
+        if g1.target_qubits != g2.target_qubits:
+            return False
+        if (
+            getattr(g1, "is_controlled_by", True)
+            and g1.control_qubits != g2.control_qubits
+        ):
+            return False
+
+        # Identity conditions
+        name = g1.name
+        if name in ("h", "cx"):
+            return True
+
+        if name == "ry":
+            theta1 = g1.parameters[0]
+            theta2 = g2.parameters[0]
+            # Check if theta1 + theta2 is a multiple of 2π
+            return bool((theta1 + theta2) % (2 * pi) < 1e-8)
```

Comment:
even here, why ``ry`` precisely? I guess you would need to have it generally for parametrized gates, right?
Something like this:
```suggestion
        if isinstance(g1, gates.ParametrizedGate):
            # Check if theta1 + theta2 is a multiple of 2π
            same_angle = [bool((theta1 + theta2) % (2 * pi) < 1e-8 for theta1, theta2 in zip(g1.parameters, g2.parameters)]
            return all(same_angle)
```

Reply:
I raised this issue before in a comment and asked for directions on how to proceed. With @renatomello, as it my understanding, we discussed that we should leave it like this in a way that is not general and wait for #1606 to be merged such that one can use the methods defined there (InverseCancellation, RotationGateFusion) to make it general.

Reply:
IIRC `ry`, `rx`, and `rz` all have the same parent class `_Rn_`, so one could check for that class.